### DISCUSSION TITLE
How can I prevent the alt text in markdown cell of a ipynb getting wrapped?

### DISCUSSION THREAD

**Discussion by jonathanalgar** [quarto-dev/quarto-cli#8131] [Q&A]
### Description

Hello folks,

I'm trying to improve the alt texts for the images of an open-source project I follow. The PR: https://github.com/langchain-ai/langchain/pull/15357/files

The problem is for the ipynb the [build](https://github.com/langchain-ai/langchain/blob/master/docs/.local_build.sh#L23) (`quarto preview docs`) is breaking because the intermediate md files are getting wrapped. For example:

```
+ quarto preview docs --verbose
Preparing to preview

[1/1] docs/guides/local_llms.ipynb
[INFO] Loaded /opt/quarto/share/pandoc/templates/default.markdown from /opt/quarto/share/pandoc/templates/default.markdown
[INFO] Running filter /opt/quarto/share/filters/main.lua
[INFO] Completed filter /opt/quarto/share/filters/main.lua in 331 ms
```

Creates an intermediate md:

```
...
![Graphical representation of open-source LLMs assessed by base model
and fine-tuning approach.](../../static/img/OSS_LLM_overview.png)
...
```

Which breaks the Docusaurus HTML build:

```
SyntaxError: /home/ja/langchain-2/_dist/docs/guides/local_llms.md: Unterminated string constant. (42:13)
  40 | was used?`}</li>
  41 | </ol>
> 42 | <p><img alt={"Graphical representation of open-source LLMs assessed by base model
     |              ^
  43 | and fine-tuning approach."}...
```

Do you have any suggestions on what I could do myself in the ipynb file or a minimal change I could ask the maintainers to do to fix?

Many thanks 😊

**Comment by mcanouil** [comment]
Could you share a small and simple self-contained "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.
There are too many files and it's an ongoing PR.

**Reply by cderv** [reply]
@jonathanalgar Especially can you edit your post and share 

* which format you are targetting ? `format: docusaurus-md` ? 
* Which Quarto version are you using ? 


I tried something like below but it seems to produce something that shows ok testing on https://docusaurus.new/

````markdown
---
title: Run LLMs locally
format: docusaurus-md
keep-md: true
---

![Graphical representation of open-source LLMs assessed by base model and fine-tuning approach.](https://raw.githubusercontent.com/jonathanalgar/langchain/update-alt-texts-using-alttexter/docs/static/img/OSS_LLM_overview.png)

````

**Reply by jonathanalgar** [reply]
@mcanouil Thank you for the speedy response. I can replicate this in a self-contained way like,
```
$ wget https://raw.githubusercontent.com/jonathanalgar/langchain/update-alt-texts-using-alttexter/docs/docs/guides/local_llms.ipynb -O local_llms.ipynb > /dev/null 2>&1
$ quarto --version
1.3.450
$ quarto render local_llms.ipynb --to markdown --verbose
pandoc --verbose
  to: markdown
  output-file: local_llms.md
  standalone: true
  default-image-extension: png

metadata
  title: Run LLMs locally

[INFO] Running filter /opt/quarto/share/filters/main.lua
[INFO] Completed filter /opt/quarto/share/filters/main.lua in 327 ms
Output created: local_llms.md
$ awk -v RS='' -v ORS='\n\n' '/!\[.*\]\(.*\)/ { print }' local_llms.md
![Graphical representation of open-source LLMs assessed by base model
and fine-tuning approach.](../../static/img/OSS_LLM_overview.png)

![Table showing memory required for LLaMa weights with different numbers
of parameters and data
types.](../../static/img/llama-memory-weights.png)

![Chart comparing tokens per second for 4-bit quantized LLaMa models
across different GPUs.](../../static/img/llama_t_put.png)
``` 

(1) Can I override this wrapping behavior at the markdown cell level in the ipynb file?
(2) What would be the minimal change that could be made to `/opt/quarto/share/filters/main.lua` to override this behavior?

Or am I missing a trick?

**Reply by jonathanalgar** [reply]
@cderv Thank you also for the speedy response.

This works for me as well. But the image image line is intact. My problem is that the image is split over two lines.
Testing outside the build (as in my post just above) I get this output with both `--to markdown` and `--to docusaurus-md`.

Edit: what does work is if I add `--wrap none` to override the [default pandoc wrap behavior](https://pandoc.org/MANUAL.html#:~:text=With%20auto%20(the%20default)%2C,not%20wrap%20lines%20at%20all.).

i.e. `quarto render local_llms.ipynb --to markdown --verbose --wrap none`

So one solution could be to propose to the maintainers to add that option to their build. But I have to try to understand if it will impact anything else.

**Reply by cderv** [reply]
Which quarto version are you using ? I think we did some changes on 1.4 and we output .mdx format.

Also, could you provide a reduce example of you ipynb with only the content that breaks ? 

Can you provide docusaurus version too you are using ? 

I am testing output on https://docusaurus.new playground and it seems to build correctly when pasting something that break line in the caption syntax  like
````mdx
![Graphical representation of open-source LLMs assessed by base model
and fine-tuning approach.](https://raw.githubusercontent.com/jonathanalgar/langchain/update-alt-texts-using-alttexter/docs/static/img/OSS_LLM_overview.png)
````

**Reply by jonathanalgar** [reply]
@cderv quarto version 1.3.450. The repo is using docusaurus 2.4.3. As per your example upgrading that to 3.0.1 fixes the issues with the alt texts in the local build and render this particular doc perfectly in HTML. Do you know exactly what change in  docusaurus fixed this? Unfortunately there are unrelated build errors for other files in the repo caused by [breaking changes](https://docusaurus.io/blog/preparing-your-site-for-docusaurus-v3#bad-usage-of--1),

```
Error: MDX compilation failed for file "/home/ja/langchain-2/_dist/docs/integrations/retrievers/self_query/supabase_self_query.md"
Cause: Unexpected character `=` (U+003D) before name, expected a character that can start a name, such as a letter, `$`, or `_`
Details:
{
  "column": 36,
  "message": "Unexpected character `=` (U+003D) before name, expected a character that can start a name, such as a letter, `$`, or `_`",
  "line": 56,
  "name": "56:36",
  "place": {
    "line": 56,
    "column": 36,
    "offset": 2288,
    "_index": 22,
    "_bufferIndex": 31
  },
  "reason": "Unexpected character `=` (U+003D) before name, expected a character that can start a name, such as a letter, `$`, or `_`",
  "ruleId": "unexpected-character",
  "source": "micromark-extension-mdx-jsx",
  "url": "https://github.com/micromark/micromark-extension-mdx-jsx#unexpected-character-at-expected-expect"
}
```

Ideally I'd like to get the alt texts working without having to propose upgrading docusaurus.

As an alternative, I thought I could propose to the maintainers to use the `wrap none` option in the quarto for the documents I've updated. Locally this works for `--to markdown`

```
$ quarto render local_llms.ipynb --to markdown --verbose --wrap none
pandoc --verbose --wrap none
  to: markdown
  output-file: local_llms.md
  standalone: true
  default-image-extension: png

metadata
  title: Run LLMs locally

[INFO] Running filter /opt/quarto/share/filters/main.lua
[INFO] Completed filter /opt/quarto/share/filters/main.lua in 322 ms
Output created: local_llms.md

$ awk -v RS='' -v ORS='\n\n' '/!\[.*\]\(.*\)/ { print }' local_llms.md
![Graphical representation of open-source LLMs assessed by base model and fine-tuning approach.](../../static/img/OSS_LLM_overview.png)

![Table showing memory required for LLaMa weights with different numbers of parameters and data types.](../../static/img/llama-memory-weights.png)

![Chart comparing tokens per second for 4-bit quantized LLaMa models across different GPUs.](../../static/img/llama_t_put.png)
```

but not for `--to docusaurus-md`:

```
$ quarto render local_llms.ipynb --to docusaurus-md --verbose --wrap none
pandoc --verbose --wrap none
  to: >-
    markdown_strict+raw_html+all_symbols_escapable+backtick_code_blocks+fenced_code_blocks+space_in_atx_header+intraword_underscores+lists_without_preceding_blankline+shortcut_reference_links+autolink_bare_uris+emoji+footnotes+gfm_auto_identifiers+pipe_tables+strikeout+task_lists+tex_math_dollars+pipe_tables+tex_math_dollars+header_attributes+raw_html+all_symbols_escapable+backtick_code_blocks+fenced_code_blocks+space_in_atx_header+intraword_underscores+lists_without_preceding_blankline+shortcut_reference_links
  output-file: local_llms.md
  standalone: true
  default-image-extension: png
  wrap: none
  html-math-method: webtex

metadata
  title: Run LLMs locally

[INFO] Loaded /opt/quarto/share/pandoc/templates/default.markdown from /opt/quarto/share/pandoc/templates/default.markdown
[INFO] Running filter /opt/quarto/share/filters/main.lua
[INFO] Completed filter /opt/quarto/share/filters/main.lua in 321 ms
Output created: local_llms.md

$ awk -v RS='' -v ORS='\n\n' '/!\[.*\]\(.*\)/ { print }' local_llms.md
![Graphical representation of open-source LLMs assessed by base model
and fine-tuning approach.](../../static/img/OSS_LLM_overview.png)

![Table showing memory required for LLaMa weights with different numbers
of parameters and data
types.](../../static/img/llama-memory-weights.png)

![Chart comparing tokens per second for 4-bit quantized LLaMa models
across different GPUs.](../../static/img/llama_t_put.png)
```

So I'm not really sure what my options are.

**Reply by mcanouil** [reply]
You can pass that option to Quarto but in the end, if they stay on an old version of Docusaurus, it's not really up to Quarto/Pandoc to fix this.
Quarto can't really be retro compatible with all third parties.
Docusaurus 2.4.3 is more recent than Quarto 1.3.450.
Could you try with Quarto 1.4 pre-release?

EDIT: the issue here seems to be that it's a custom rendered that does not take into account `wrap`.
For instance, if you add the following in your Jupyter Notebook, you'll see an intermediate markdown file without wrapping while it is wrapped in the mdx file.
I am not very familiar with that format and that part of the codebase unfortunately.

````qmd
---
wrap: none
keep-md: true
---
````


FYI, the Notebook is very well hidden in all the code blocks.
Here it is: https://github.com/jonathanalgar/langchain/blob/update-alt-texts-using-alttexter/docs/docs/guides/local_llms.ipynb

**Reply by cderv** [reply]
> EDIT: the issue here seems to be that it's a custom rendered that does not take into account wrap.

You mean `docusaurus-md` not respecting `--wrap` ? This is what I think too and is highly possibly. 

> Ideally I'd like to get the alt texts working without having to propose upgrading docusaurus.

You should definitely follow advice and try Quarto 1.4 pre-release with your docusaurus version. I think we are using `<figure>` tag in 1.4. 

Can you try that ? 

Otherwise other workaround will require Lua filter override or writing raw docusuarus in your input

**Reply by jonathanalgar** [reply]
Yes, thanks, moving to Quarto v1.4.534 fixes the issue with the wrapping by using `<figure>` tag. And across the entire repo the render to docusaurus-md with the new version fails with just with one file:

**[rockset.ipynb](https://github.com/jonathanalgar/langchain/blob/master/docs/docs/integrations/vectorstores/rockset.ipynb):**
```
Error running filter /opt/quarto/share/filters/main.lua:
...are/extensions/quarto/docusaurus/docusaurus_citeproc.lua:8: attempt to index a nil value (field 'integer index')
stack traceback:
        [C]: in ?
        [C]: in method 'walk'
        ...are/extensions/quarto/docusaurus/docusaurus_citeproc.lua:4: in function <...are/extensions/quarto/docusaurus/docusaurus_citeproc.lua:1>
        [C]: in ?
        [C]: in method 'walk'
        /opt/quarto/share/filters/main.lua:224: in local 'checked_walk'
        /opt/quarto/share/filters/main.lua:276: in function 'run_emulated_filter'
        /opt/quarto/share/filters/main.lua:936: in local 'callback'
        [string "..."]:1888: in field 'withScriptFile'
        /opt/quarto/share/filters/main.lua:949: in upvalue 'run_emulated_filter_chain'
        /opt/quarto/share/filters/main.lua:987: in function </opt/quarto/share/filters/main.lua:984>
stack traceback:
        ...are/extensions/quarto/docusaurus/docusaurus_citeproc.lua:4: in function <...are/extensions/quarto/docusaurus/docusaurus_citeproc.lua:1>
        [C]: in ?
        [C]: in method 'walk'
        /opt/quarto/share/filters/main.lua:224: in local 'checked_walk'
        /opt/quarto/share/filters/main.lua:276: in function 'run_emulated_filter'
        /opt/quarto/share/filters/main.lua:936: in local 'callback'
        [string "..."]:1888: in field 'withScriptFile'
        /opt/quarto/share/filters/main.lua:949: in upvalue 'run_emulated_filter_chain'
        /opt/quarto/share/filters/main.lua:987: in function </opt/quarto/share/filters/main.lua:984>
stack traceback:
        /opt/quarto/share/filters/main.lua:224: in local 'checked_walk'
        /opt/quarto/share/filters/main.lua:276: in function 'run_emulated_filter'
        /opt/quarto/share/filters/main.lua:936: in local 'callback'
        [string "..."]:1888: in field 'withScriptFile'
        /opt/quarto/share/filters/main.lua:949: in upvalue 'run_emulated_filter_chain'
        /opt/quarto/share/filters/main.lua:987: in function </opt/quarto/share/filters/main.lua:984>
```
Resolved by fixing the formatting error:
https://github.com/jonathanalgar/langchain/blob/master/docs/docs/integrations/vectorstores/rockset.ipynb?short_path=8a0406f#L20
—is this a bug or the expected behavior?

Though quickly ran into a problem when testing on the docusaurus server (2.4.3) for **[local_llms.ipynb](https://github.com/jonathanalgar/langchain/blob/update-alt-texts-using-alttexter/docs/docs/guides/local_llms.ipynb).**. the image doesn't render:

![chrome_htodnlCE59](https://github.com/quarto-dev/quarto-cli/assets/93204286/2cf48bdb-3297-4757-ac4c-823075fab863)

**local_llms.mdx:**
```
...
<figure>
<img src="../../static/img/OSS_LLM_overview.png"
alt="Graphical representation of open-source LLMs assessed by base model and fine-tuning approach." />
<figcaption aria-hidden="true">Graphical representation of open-source
LLMs assessed by base model and fine-tuning approach.</figcaption>
</figure>
...
```

**Reply by cderv** [reply]
> Resolved by fixing the formatting error:
[jonathanalgar/langchain@master/docs/docs/integrations/vectorstores/rockset.ipynb?short_path=8a0406f#L20](https://github.com/jonathanalgar/langchain/blob/master/docs/docs/integrations/vectorstores/rockset.ipynb?short_path=8a0406f#L20)
—is this a bug or the expected behavior?

It seems indeed empty link are creating an issue... 
````markdown
---
title: test
format: docusaurus-md
---

[ ](https://python.langchain.com/docs/modules/data_connection/vectorstores/integrations/rockset#setting-up-environment)

````

<details>
<summary>Error</summary>

````
Error running filter C:\Users\chris\DOCUME~1\DEV_R\QUARTO~2\src\resources\filters\main.lua:
...ces\extensions\quarto\docusaurus\docusaurus_citeproc.lua:8: attempt to index a nil value (field 'integer index')
stack traceback:
	[C]: in ?
	[C]: in method 'walk'
	...ces\extensions\quarto\docusaurus\docusaurus_citeproc.lua:4: in function <...ces\extensions\quarto\docusaurus\docusaurus_citeproc.lua:1>
	[C]: in ?
	[C]: in method 'walk'
	...V_R\QUARTO~2\src\resources\filters\./ast/customnodes.lua:86: in local 'checked_walk'
	...V_R\QUARTO~2\src\resources\filters\./ast/customnodes.lua:138: in function 'run_emulated_filter'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:82: in local 'callback'
	[string "..."]:1888: in field 'withScriptFile'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:95: in upvalue 'run_emulated_filter_chain'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:133: in function <..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:130>
stack traceback:
	...ces\extensions\quarto\docusaurus\docusaurus_citeproc.lua:4: in function <...ces\extensions\quarto\docusaurus\docusaurus_citeproc.lua:1>
	[C]: in ?
	[C]: in method 'walk'
	...V_R\QUARTO~2\src\resources\filters\./ast/customnodes.lua:86: in local 'checked_walk'
	...V_R\QUARTO~2\src\resources\filters\./ast/customnodes.lua:138: in function 'run_emulated_filter'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:82: in local 'callback'
	[string "..."]:1888: in field 'withScriptFile'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:95: in upvalue 'run_emulated_filter_chain'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:133: in function <..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:130>
stack traceback:
	...V_R\QUARTO~2\src\resources\filters\./ast/customnodes.lua:86: in local 'checked_walk'
	...V_R\QUARTO~2\src\resources\filters\./ast/customnodes.lua:138: in function 'run_emulated_filter'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:82: in local 'callback'
	[string "..."]:1888: in field 'withScriptFile'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:95: in upvalue 'run_emulated_filter_chain'
	..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:133: in function <..._R\QUARTO~2\src\resources\filters\./ast/runemulation.lua:130>
ERROR: Error
    at renderFiles (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/command/render/render-files.ts:346:23)
    at eventLoopTick (ext:core/01_core.js:183:11)
    at async renderProject (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/command/render/project.ts:268:23)
    at async renderForPreview (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/command/preview/preview.ts:419:24)
    at async render (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/command/preview/preview.ts:165:22)
    at async preview (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/command/preview/preview.ts:182:18)
    at async Command.fn (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/command/preview/cmd.ts:415:7)
    at async Command.execute (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/vendor/deno.land/x/cliffy@v0.25.4/command/command.ts:1790:7)
    at async quarto (file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/quarto.ts:155:3)
    at async file:///C:/Users/chris/DOCUME~1/DEV_R/QUARTO~2/src/quarto.ts:186:5
````

</details>

It should not create this type of Lua error, though empty link syntax is uncommon... 

What is the expected output to add a link to empty string ?

**Reply by cderv** [reply]
I fix it in https://github.com/quarto-dev/quarto-cli/issues/8159 by the way, but still unsure what is the expected output in your case.

**Reply by jonathanalgar** [reply]
> I fix it in https://github.com/quarto-dev/quarto-cli/issues/8159 by the way, but still unsure what is the expected output in your case.

Correct, it is an error in the content that I will propose a correction for. I just wanted to point out the error being thrown.

With regards to the images not rendering correctly for the `<figure>` tag on the docusaurus 2.4.3, this makes me think there is a backwards compatibility issue and I should wait to update the alt texts on the ipynb until the docusaurus for the project is updated to v3.
